### PR TITLE
Parameter attributes

### LINKED ISSUE #5032: Support attributes on parameters
Dafny currently does not support attributes on parameters. Allowing attributes here is useful for custom back-ends that may wish to annotate parameters to customize code generation

Example:
```dafny
function Foo({:should-be-possible} x: int): int {
  3
}

datatype A = A({:should-be-possible} x: int)
```

### PR DESCRIPTION
**Pull request by keyboardDrummer**
Fixes #5032

### Description
Support attributes in parameters

### How has this been tested?
Updated an existing littish test

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: docs/dev/news/5032.feat [hunk 1]
```diff
@@ -0,0 +1 @@
+Function and method parameters and return types, and datatype constructor arguments, can now have attributes. By default, there are no attributes that Dafny recognises in these positions, but custom back-ends can use this feature to get extra information from the source files.
```
Comment (MikaelMayer) [inline docs/dev/news/5032.feat:1]:
```suggestion
Function and method parameters and return types, and datatype constructor arguments, can now have attributes. By default, there are no attributes that Dafny recognizes in these positions, but custom back-ends can use this feature to get extra information from the source files.
```

File: Source/DafnyCore/Dafny.atg [hunk 2]
```diff
@@ -964,16 +964,19 @@ IdentTypeOptional<out BoundVar var>
...
      Contract.Ensures(Contract.ValueAtReturn(out ty)!=null);
      Contract.Ensures(Contract.ValueAtReturn(out identName)!=null);
      IToken nameToken = null; ty = new BoolType()/*dummy*/; isGhost = false;
      IToken nameonlyToken = null;
+     attributes = null;
      identName = null;
      defaultValue = null;
   .)
   (. var beforeStartToken = t; .)
+  { Attribute<ref attributes> }
```
Comment (MikaelMayer) [inline Source/DafnyCore/Dafny.atg:979]: Indeed, I validate putting these before the other modifiers like ghost or older. We should do the same and put attributes before function and methods keywords as well, but let's keep it for another time.

**Review by MikaelMayer — CHANGES_REQUESTED** [review]
Good nice to have. Could you also update the documentation of attributes to add that attributes are supported on parameters now?

**Comment by MikaelMayer** [comment]
> As far as I see, the documentation doesn't say what types of AST nodes support attributes, so it doesn't seem to make sense to say just for parameters that they support attributes.

Can you look for the following in https://dafny.org/latest/DafnyRef/DafnyRef ?
`Formals(allowGhostKeyword, allowNewKeyword`
I think you might want to add `{ Attribute }` before the `GIdentType`

**Review by MikaelMayer — APPROVED** [review]

**Comment by RustanLeino** [comment]
@keyboardDrummer Sorry to be late to the party, but would you mind adding some tests to `dafny0/AttributeChecks.dfy`. The tests in that file try to make sure that all processing that needs to happen with attributes (like type checking, making sure type characteristics are used properly, etc.) happen.

**Pull request merged**